File: Day 19/solution.cpp
#include <filesystem>
#include <sstream>
#include <fstream>
#include <iostream>
#include <vector>
#include <functional>
#include <unordered_map>

typedef std::unordered_map<std::string, std::vector<std::tuple<char, char, int, std::string>>> Workflow;

class Solver
{
	std::filesystem::path filepath;

	static int variable_name_to_idx(char c)
	{
		switch (c)
		{
		case 'x':
			return 0;
		case 'm':
			return 1;
		case 'a':
			return 2;
		case 's':
			return 3;
		default:
			throw "Invalid variable name.";
		}
	};

	long generic_solve(const std::function<long(const Workflow &, std::ifstream &)> &solve)
	{
		std::ifstream file(this->filepath);
		Workflow workflow;
		for (char c = file.get(); c != '\n'; c = file.get())
		{
			std::string workflow_name;
			for (; c != '{'; c = file.get())
				workflow_name += c;

			// Get all the conditions
			while (c != '}')
			{
				c = file.get();

				// Get one condition
				bool colon_found = false;
				std::string variable_name, target;
				char operation = 0;
				int value = 0;
				for (; c != ',' && c != '}'; c = file.get())
				{
					if (c == '>' || c == '<')
					{
						operation = c;
					}
					else if (c == ':')
					{
						colon_found = true;
					}
					else if (colon_found)
					{
						target += c;
					}
					else if (operation != 0)
					{
						value = value * 10 + c - '0';
					}
					else
					{
						variable_name += c;
					}
				}

				if (operation == 0)
				{
					std::swap(target, variable_name);
				}

				workflow[workflow_name].emplace_back(variable_name[0], operation, value, target);
			}

			file.get(); // Ignore new line
		}

		return solve(workflow, file);
	}

	long part_1()
	{
		auto solve = [](const Workflow &workflow, std::ifstream &file) -> long
		{
			long total = 0;
			auto operations = [](int a, char operation, int b)
			{
				switch (operation)
				{
				case '>':
					return a > b;
				case '<':
					return a < b;
				default:
					throw "Invalid operation.";
				}
			};

			for (char c = file.get(); c != EOF; c = file.get())
			{
				int values[4];
				// Get all statements
				for (c = file.get(); c != '\n' && c != EOF; c = file.get())
				{
					// Get one statement
					int var_index = -1, value = 0;
					for (; c != ',' && c != '}'; c = file.get())
					{
						if (c == '=')
							continue;

						if (var_index == -1)
						{
							var_index = variable_name_to_idx(c);
							continue;
						}

						value = value * 10 + c - '0';
					}
					values[var_index] = value;
				}

				std::string current = "in";
				while (current != "A" && current != "R")
				{
					for (auto [variable_name,
										 operation,
										 value,
										 target] : workflow.at(current))
					{
						if ((operation == 0) ||
								operations(
										values[Solver::variable_name_to_idx(variable_name)],
										operation,
										value))
						{
							current = target;
							break;
						}
					}
				}

				if (current == "A")
				{
					for (int i = 0; i < 4; ++i)
					{
						total += values[i];
					}
				}
			}

			return total;
		};

		return this->generic_solve(solve);
	}

	long part_2()
	{
		auto get_new_ranges = [](
															int low,
															int high,
															int value,
															char operation) -> std::pair<std::pair<int, int>, std::pair<int, int>>
		{
			switch (operation)
			{
			case '<':
				return {{low, value - 1}, {value, high}};
			case '>':
				return {{value + 1, high}, {low, value}};
			default:
				throw "Invalid operation.";
			}
		};

		auto count_unique_arrangements = [get_new_ranges](
																				 auto self,
																				 const Workflow &workflow,
																				 const std::string &current,
																				 std::pair<int, int>(&ranges)[4]) -> long
		{
			if (current == "R")
				return 0;

			if (current == "A")
			{
				long result = 1;
				for (int i = 0; i < 4; ++i)
				{
					result *= ranges[i].second - ranges[i].first + 1;
				}
				return result;
			}

			long result = 0;
			std::pair<int, int> original_ranges[4];
			for (int i = 0; i < 4; ++i)
			{
				original_ranges[i] = ranges[i];
			}

			for (auto [variable_name, operation, value, target] : workflow.at(current))
			{
				if (operation == 0)
				{
					result += self(self, workflow, target, ranges);
					continue;
				}

				int variable_idx = Solver::variable_name_to_idx(variable_name);
				auto [low, high] = ranges[variable_idx];
				auto [positive_range, negative_range] = get_new_ranges(low, high, value, operation);
				if (positive_range.first > positive_range.second ||
						negative_range.first > negative_range.second)
					break;

				ranges[variable_idx] = positive_range;
				result += self(self, workflow, target, ranges);
				ranges[variable_idx] = negative_range;
			}

			for (int i = 0; i < 4; ++i)
			{
				ranges[i] = original_ranges[i];
			}
			return result;
		};

		auto solve = [count_unique_arrangements](
										 const Workflow &workflow,
										 std::ifstream &file) -> long
		{
			file.close();

			std::pair<int, int> ranges[4]{{1, 4000}, {1, 4000}, {1, 4000}, {1, 4000}};
			return count_unique_arrangements(count_unique_arrangements, workflow, "in", ranges);
		};
		return this->generic_solve(solve);
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 19" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}

File: Day 19/solution.py
"""
Day 19 solution.
"""
import sys
from io import TextIOWrapper
from typing import Callable, Iterable

Workflow = dict[str, list[tuple[str, str, int, str]]]


class Solver:
    """Day 19 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    @staticmethod
    def product(nums: Iterable[int]) -> int:
        """Product of a sequence of numbers."""
        result = 1
        for x in nums:
            result *= x
        return result

    def generic_solve(
        self, solve: Callable[[Workflow, TextIOWrapper], int]
    ) -> int:
        """Generic solve."""
        workflow: Workflow = {}
        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            for line in file:
                if line == "\n":
                    break

                workflow_name, conditions = line.strip()[:-1].split("{")
                workflow[workflow_name] = []
                for condition_statement in conditions.split(","):
                    if ":" not in condition_statement:
                        workflow[workflow_name].append(
                            ("", "", 0, condition_statement)
                        )
                        continue

                    condition, target = condition_statement.split(":")
                    variable_name, operation, value = (
                        condition[0],
                        condition[1],
                        condition[2:],
                    )
                    workflow[workflow_name].append(
                        (variable_name, operation, int(value), target)
                    )

            return solve(workflow, file)

    def part_1(self) -> int:
        """Part 1 solver."""

        def solve(workflow: Workflow, file: TextIOWrapper) -> int:
            total = 0
            operations = {">": lambda a, b: a > b, "<": lambda a, b: a < b}
            for line in file:
                values = {}
                for statements in line.strip()[1:-1].split(","):
                    variable_name, value = statements.split("=")
                    values[variable_name] = int(value)

                current = "in"
                while current not in ["A", "R"]:
                    for variable_name, operation, value, target in workflow[
                        current
                    ]:
                        if (operation == "") or operations[operation](
                            values[variable_name], value
                        ):
                            current = target
                            break

                if current == "A":
                    total += sum(values.values())

            return total

        return self.generic_solve(solve)

    def part_2(self) -> int:
        """Part 2 solver."""

        def solve(workflow: Workflow, _: TextIOWrapper) -> int:
            def get_new_ranges(
                low: int, high: int, value: int, operation: str
            ) -> tuple[tuple[int, int], tuple[int, int]]:
                if operation == "<":
                    return (low, value - 1), (value, high)
                if operation == ">":
                    return (value + 1, high), (low, value)
                raise ValueError("Invalid operation.")

            def count_unique_arrangements(
                current: str, ranges: dict[str, tuple[int, int]]
            ) -> int:
                if current == "R":
                    return 0

                if current == "A":
                    return self.product(
                        high - low + 1 for low, high in ranges.values()
                    )

                original_ranges = dict(ranges)
                result = 0
                for variable_name, operation, value, target in workflow[
                    current
                ]:
                    if operation == "":
                        result += count_unique_arrangements(target, ranges)
                        continue

                    next_ranges = (
                        positive_range,
                        negative_range,
                    ) = get_new_ranges(
                        *ranges[variable_name], value, operation
                    )
                    if any(low > high for low, high in next_ranges):
                        break

                    ranges[variable_name] = positive_range
                    result += count_unique_arrangements(target, ranges)
                    ranges[variable_name] = negative_range

                for variable_name, range in original_ranges.items():
                    ranges[variable_name] = range
                return result

            return count_unique_arrangements(
                "in", {variable_name: (1, 4000) for variable_name in "xmas"}
            )

        return self.generic_solve(solve)

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 19")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case _:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()
